gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/service/InboundMessageMergingService.java

```diff
@@ -0,0 +1,162 @@
+    private final MigrationStatusLogService migrationStatusLogService;
+    private final XmlParseUtilService xmlParseUtilService;
+    private final ObjectMapper objectMapper;
+    private final XPathService xPathService;
+    private final FhirParser fhirParser;
+    private final BundleMapperService bundleMapperService;
+    private final PatientMigrationRequestDao migrationRequestDao;
+    private final NackAckPreparationService nackAckPreparationService;
+    private static final String CONVERSATION_ID_HAS_NOT_BEEN_GIVEN = "Conversation Id has not been given";
+
+    public boolean canMergeCompleteBundle(String conversationId) throws ValidationException {
+
+        if (!StringUtils.hasText(conversationId)) {
+            throw new ValidationException(CONVERSATION_ID_HAS_NOT_BEEN_GIVEN);
+        }
+
+        var undeletedLogs = getUndeletedLogsForConversation(conversationId);
+        return undeletedLogs.stream().allMatch(log -> log.getUploaded().equals(true));
+    }
+
+    public void mergeAndBundleMessage(String conversationId) throws JAXBException, JsonProcessingException {
+
+        if (!StringUtils.hasText(conversationId)) {
+            throw new ValidationException(CONVERSATION_ID_HAS_NOT_BEEN_GIVEN);
+        }
+
+        PatientMigrationRequest migrationRequest = migrationRequestDao.getMigrationRequest(conversationId);
+        var inboundMessage = objectMapper.readValue(migrationRequest.getInboundMessage(), InboundMessage.class);
+        RCMRIN030000UK06Message payload = unmarshallString(inboundMessage.getPayload(), RCMRIN030000UK06Message.class);
+
+        try {
+
+            var attachmentLogs = getUndeletedLogsForConversation(conversationId);
+            var attachmentsContainSkeletonMessage = attachmentLogs.stream().anyMatch(log -> log.getSkeleton().equals(true));
+
+            if (attachmentsContainSkeletonMessage) {
+                // merge skeleton message into original payload
+                var skeletonLogs = attachmentLogs.stream().filter(log -> log.getSkeleton().equals(true)).toList();
+                var skeletonFileName = skeletonLogs.stream().findFirst().get().getFilename();
+                var skeletonFileAsString = new String(attachmentHandlerService.getAttachment(skeletonFileName), StandardCharsets.UTF_8);
+
+                // get ebxml references to find document id from skeleton message
+                List<EbxmlReference> attachmentReferenceDescription = new ArrayList<>();
+                attachmentReferenceDescription.addAll(xmlParseUtilService.getEbxmlAttachmentsData(inboundMessage));
+                var ebxmlSkeletonReference = attachmentReferenceDescription
+                        .stream()
+                        .filter(reference -> reference.getHref().contains(skeletonLogs.get(0).getMid()))
+                        .findFirst();
+
+                var skeletonDocumentId = ebxmlSkeletonReference.get().getDocumentId();
```

Comment:
Is it possible that the `ebxmlSkeletonReference` is not found in the `attachmentReferenceDescription`? If yes, then you should check if `ebxmlSkeletonReference.isPresent()` before calling `get()` method.